pkg/tbtc/signing.go

```diff
@@ -2,378 +2,386 @@ package tbtc
-// their controlled signing group members.
-func (srl *signingRetryLoop) qualifiedOperatorsSet(
-	readyMembersIndexes []group.MemberIndex,
-) (map[chain.Address]bool, error) {
-	// The retry algorithm expects that we count retries from 0. Since
-	// the first invocation of the algorithm will be for `attemptCounter == 1`
-	// we need to subtract one while determining the number of the given retry.
-	retryCount := srl.attemptCounter - 1
-
-	var readySigningGroupOperators []chain.Address
-	for _, memberIndex := range readyMembersIndexes {
-		readySigningGroupOperators = append(
-			readySigningGroupOperators,
-			srl.signingGroupOperators[memberIndex-1],
-		)
-	}
-
-	qualifiedOperators, err := retry.EvaluateRetryParticipantsForSigning(
-		readySigningGroupOperators,
-		srl.attemptSeed,
-		retryCount,
-		uint(srl.chainConfig.HonestThreshold),
+	signingLogger := logger.With(
+		zap.String("wallet", fmt.Sprintf("0x%x", walletPublicKeyBytes)),
+		zap.String("message", fmt.Sprintf("0x%x", message)),
+		zap.Uint64("signingStartBlock", startBlock),
+		zap.Uint64("signingTimeoutBlock", loopTimeoutBlock),
 	)
-	if err != nil {
-		return nil, fmt.Errorf(
-			"random operator selection failed: [%w]",
-			err,
-		)
+
+	type signerOutcome struct {
+		signature *tecdsa.Signature
+		endBlock  uint64
+		err       error
 	}
 
-	return chain.Addresses(qualifiedOperators).Set(), nil
-}
+	signerOutcomesChan := make(chan *signerOutcome, len(se.signers))
+
+	for _, currentSigner := range se.signers {
+		go func(signer *signer) {
+			se.protocolLatch.Lock()
+			defer se.protocolLatch.Unlock()
 
-// excludedMembersIndexes returns a list of excluded members' indexes for
-// the given qualified operators set.
-func (srl *signingRetryLoop) excludedMembersIndexes(
-	qualifiedOperatorsSet map[chain.Address]bool,
-	readyMembersIndexes []group.MemberIndex,
-) []group.MemberIndex {
-	includedMembersIndexes := make([]group.MemberIndex, 0)
-	excludedMembersIndexes := make([]group.MemberIndex, 0)
-	for i, operator := range srl.signingGroupOperators {
-		memberIndex := group.MemberIndex(i + 1)
-
-		if qualifiedOperatorsSet[operator] &&
-			slices.Contains(readyMembersIndexes, memberIndex) {
-			includedMembersIndexes = append(
-				includedMembersIndexes,
-				memberIndex,
+			announcer := announcer.New(
+				fmt.Sprintf("%v-%v", ProtocolName, "signing"),
+				se.chainConfig.GroupSize,
+				se.broadcastChannel,
+				se.membershipValidator,
 			)
-		} else {
-			excludedMembersIndexes = append(
-				excludedMembersIndexes,
-				memberIndex,
+
+			doneCheck := newSigningDoneCheck(
+				se.chainConfig.GroupSize,
+				se.broadcastChannel,
+				se.membershipValidator,
 			)
-		}
+
+			retryLoop := newSigningRetryLoop(
+				signingLogger,
+				message,
+				startBlock,
+				signer.signingGroupMemberIndex,
+				wallet.signingGroupOperators,
+				se.chainConfig,
+				announcer,
+				doneCheck,
+			)
+
+			// Set up the loop timeout signal.
```

Comment:
```suggestion
			// Set up the loop timeout signal. This context is associated with
			// all attempts and gets canceled in three situations:
			// - one of the attempts failed with an error,
			// - we gave up after executing multiple attempts,
			// - one of the attempts succeeded.
			// In the last case, the context is not canceled immediately but
			// we wait until the timeout for the entire loop is done. This lets
			// the inner attempt-scope context to retransmit messages for a
			// longer period of time than just for the execution time of the
			// given attempt from the perspective of the current member. This
			// is important to ensure everyone has a chance to receive
			// signingDoneMessage - see attemptCtx docs, please.
```

Reply:
The proposed comment is not entirely true (same reasons as in https://github.com/keep-network/keep-core/pull/3404#discussion_r1030369237). See my proposition 761219ec